### PR TITLE
fix(legacy): resolves the issue of incorrect signature generation when the Stellar transaction source account differs from the signing account.

### LINKED ISSUE #3690: Invalid Signature when using signer key to sign the transaction source on Stellar
**Describe the bug**
I encountered an issue with Trezor Connect while attempting to sign a Stellar transaction using a Trezor One. The source wallet is generated from Trezor, and it contains a signer which is also generated from Trezor. The process involves utilizing the TrezorConnect.stellarSignTransaction method to initiate the transaction signing with the signer. Despite successfully invoking the signing method, the process results in an "invalid signature" error upon adding the signature to the transaction.

**Info:**

-   Browser: Chrome, 123.0.6312.87
-   Browser Installation Method: downloaded installer
-   OS: Macos 14.4.1
-   Bridge: version: 2.0.27
-   Firmware Version: Trezor one, firmware version 1.12.1
-  Trezor connect web npm version: 9.2.2
-  Trezor  connect stellar plugin version: 9.0.3
-  Stellar sdk version: 11.3.0

**How to reproduce**
1. Use TrezorConnect.stellarGetAddress to get an account. Use Stellar SDK with Friendbot to fund the account. This will be the source wallet.
2. Use Stellar SDK with TransactionBuilder to build a transaction with a payment Operation.
3. Use TrezorConnect.stellarGetAddress to get a different account, it will be used as the signer.
4.  Use Stellar SDK with TransactionBuilder to build a transaction with a setOption operation, set the following:
    - min, medium, max threshold to 1
    - master key weight to 0
    - the signer address with weight 1
5. Use transformTrezorTransaction from Trezor Stellar Plugin to transform the transaction by using the path of the source wallet. Use TrezorConnect.stellarSignTransaction to sign the transformed transaction. Add the signature and submit it to the testnet. This adds the signer into the source wallet.
6. Use transformTrezorTransaction to transform the payment transaction from step 2 by using the path of the signer. Use TrezorConnect.stellarSignTransaction to sign the transformed transaction. 

Here is the code:
```
   // get an account for the source wallet
    const connectWallet = async () => {
      let address = "";
      try {
        const result = await TrezorConnect.stellarGetAddress({
          path: "m/44'/148'/" + selectedNumber + '\'',
          showOnTrezor: false,
          useEmptyPassphrase: true
        });

        if (result.success) {
          address = result.payload.address;
          console.log('Stellar Address:', address);
          setXlmAddress(address);
          setIsConnected(true);
          server.accounts().accountId(address)
          .call()
          .then((response)=> {
            console.log(response);
            setWalletBalance(response.balances[0].balance);
          })
          .catch((error)=> {
            server.friendbot(address).call()
            .then((response)=> {
              console.log(response);
            })
          });
        } else {
          console.error('Error:', result.payload.error);
        }
      } catch (error) {
        console.error('Error:', error);
      }
    }
    // generate payment transaction
    const generate = async () => {
      const query = await server.accounts().accountId(xlmAddress).call();
      const senderAccount = new Account(xlmAddress, query.sequence);

      const tx = new TransactionBuilder(senderAccount, {
        fee: "100",
        networkPassphrase: "Test SDF Network ; September 2015",
      })
      .addOperation(Operation.payment({
        source: xlmAddress,
        destination: 'GA3UK3JHOYYD3TAUH5C7NDOUDWBRF5FC4MECXFA2VRPEHIDQUOJIVOAJ',
        asset: Asset.native(),
        amount: '1'
      }))
      .setTimeout(30)
      .build();
      setUnsignedXdr(tx.toXDR());
    }

    const addSigner = async () => {
      const result = await TrezorConnect.stellarGetAddress({
        path: "m/44'/148'/25'",
        showOnTrezor: true,
        useEmptyPassphrase: true
      });

      const query = await server.accounts().accountId(xlmAddress).call();
      const senderAccount = new Account(xlmAddress, query.sequence);

      const tx = new TransactionBuilder(senderAccount, {
        fee: "100",
        networkPassphrase: "Test SDF Network ; September 2015",
      })
      .addOperation(Operation.setOptions({
        masterWeight: 0,
        lowThreshold: 1,
        medThreshold: 1,
        highThreshold: 1,
        signer: {
          ed25519PublicKey: result.payload.address,
          weight: 1
        }
      }))
      .setTimeout(30)
      .build();

      const transformed_tx = transformTrezorTransaction("m/44'/148'/25'", tx);
      const signeresult = await TrezorConnect.stellarSignTransaction(transformed_tx);
      console.log(signeresult)
      if (signeresult.success) {
        tx.addSignature(xlmAddress, Buffer.from(
          signeresult.payload.signature, "hex").toString("base64"));
      }
      server.submitTransaction(tx).then((response) => {
        console.log(response);
      });
    }

    function validate(publicKey, signature, data) {
      console.log("Public Key:", publicKey)
      console.log("Signature:", signature)
      console.log("Data:", data)

      const keypair = Keypair.fromPublicKey(publicKey)
      const verified = keypair.verify(data, Buffer.from(signature, "base64"));
      console.log("Signature Verified:", verified)
    }

    const signPayment = async () => {
      const transactionEnvelope = xdr.TransactionEnvelope.fromXDR(unSignedXdr, 'base64');
      const tx = TransactionBuilder.fromXDR(transactionEnvelope, "Test SDF Network ; September 2015");
      const transformed_tx = transformTrezorTransaction("m/44'/148'/" + selectedNumber + '\'', tx);

      const result = await TrezorConnect.stellarSignTransaction(transformed_tx);
      if (result.success) {
        const keypair = Keypair.fromPublicKey(StrKey.encodeEd25519PublicKey(Buffer.from(result.payload.publicKey, "hex")));
        const hint = keypair.signatureHint();
        const verified = keypair.verify(data, Buffer.from(result.payload.signature, "base64"));
        console.log("Signature Verified:", verified)
        tx.addDecoratedSignature(new xdr.DecoratedSignature({
          hint: hint,
          signature: Buffer.from(result.payload.signature, "hex")
        }));
      }
      setXdr3(tx.toXDR());
    }
```
### Example
Source wallet: GARO32RIIDX6EED6JKAMIWUK4CJGIRJVY7ALWIJ22DVI6TVX2NJYPDUS
Signer: GDZBFPH3APQMXMSXH7RBNKCABQFGZ7VLBO5AJ4QSDD7O3472VMT5VTXP
Generated Unsigned XDR: ```AAAAAgAAAAAi7eooQO/iEH5KgMRaiuCSZEU1x8C7ITrQ6o9Ot9NThwAAAGQADmo5AAAAAgAAAAEAAAAAAAAAAAAAAABmDjAOAAAAAAAAAAEAAAABAAAAACLt6ihA7+IQfkqAxFqK4JJkRTXHwLshOtDqj06301OHAAAAAQAAAAA3RW0ndjA9zBQ/RfaN1B2DEvSi4wgrlBqsXkOgcKOSigAAAAAAAAAAAJiWgAAAAAAAAAAA```
Invalid Signer signed XDR:
```AAAAAgAAAAAi7eooQO/iEH5KgMRaiuCSZEU1x8C7ITrQ6o9Ot9NThwAAAGQADmo5AAAAAgAAAAEAAAAAAAAAAAAAAABmDjAOAAAAAAAAAAEAAAABAAAAACLt6ihA7+IQfkqAxFqK4JJkRTXHwLshOtDqj06301OHAAAAAQAAAAA3RW0ndjA9zBQ/RfaN1B2DEvSi4wgrlBqsXkOgcKOSigAAAAAAAAAAAJiWgAAAAAAAAAAB+qsn2gAAAEAnDRt8rDMno0fuYAnEnaGtXxfuEwPWtZ9bqO3MHGEblky1MzwE52203xJPfGQuY/9K3F/NQxqER1bfZOEiENIF```
Screenshot 2024-04-04 at 3 45 25 PM

**Expected behavior**
Trezor always generates valid signature for the input transaction

**Additional context**
#### Signing with master key or a KeyPair Signer works
The signature is valid when signing with the master key, continued from the example above,
Valid XDR signed by GARO32RIIDX6EED6JKAMIWUK4CJGIRJVY7ALWIJ22DVI6TVX2NJYPDUS: ```AAAAAgAAAAAi7eooQO/iEH5KgMRaiuCSZEU1x8C7ITrQ6o9Ot9NThwAAAGQADmo5AAAAAgAAAAEAAAAAAAAAAAAAAABmDjAOAAAAAAAAAAEAAAABAAAAACLt6ihA7+IQfkqAxFqK4JJkRTXHwLshOtDqj06301OHAAAAAQAAAAA3RW0ndjA9zBQ/RfaN1B2DEvSi4wgrlBqsXkOgcKOSigAAAAAAAAAAAJiWgAAAAAAAAAABt9NThwAAAEApT+QxVD0mj8VnX3ZtZ5v1Bjv4D7K7cIGuE/F9bNJ9/zPPGLK8DKs1GAJwAiHOb/bL7pDHNFWI/Gad8D0t6d4B```
Screenshot 2024-04-04 at 3 48 26 PM

The signature is valid when signing with a KeyPair signer
Source Wallet: GDQA7QXCACLZWIJUIBHGBFNTOXDFDH3FHIK47BB3WU3OOWDXEXMOUEZV
Signer Public Key: GAKTLJG7O5YWPUMRTH3L42NY3655FDYXLHAN346IEMTGGCBJTQBVO5PV
Signer Secret Key: SDAZQA3WYLOQWENENXYKGJ4E6ZOR3YC5KOLJZNJQR4UMSHOZEBESHBIL
Unsigned XDR: ```AAAAAgAAAADgD8LiAJebITRATmCVs3XGUZ9lOhXPhDu1NudYdyXY6gAAAGQADmsVAAAAAQAAAAEAAAAAAAAAAAAAAABmDjR3AAAAAAAAAAEAAAABAAAAAOAPwuIAl5shNEBOYJWzdcZRn2U6Fc+EO7U251h3JdjqAAAAAQAAAAA3RW0ndjA9zBQ/RfaN1B2DEvSi4wgrlBqsXkOgcKOSigAAAAAAAAAAAJiWgAAAAAAAAAAA```
Signed XDR:
```AAAAAgAAAADgD8LiAJebITRATmCVs3XGUZ9lOhXPhDu1NudYdyXY6gAAAGQADmsVAAAAAQAAAAEAAAAAAAAAAAAAAABmDjR3AAAAAAAAAAEAAAABAAAAAOAPwuIAl5shNEBOYJWzdcZRn2U6Fc+EO7U251h3JdjqAAAAAQAAAAA3RW0ndjA9zBQ/RfaN1B2DEvSi4wgrlBqsXkOgcKOSigAAAAAAAAAAAJiWgAAAAAAAAAABKZwDVwAAAEA2XdAFNKzVyPQH9qJx8xSp1byDyYmWNRrWi2AV1OyeKlc1RO3rs6to8awIC58hsBxFZLrAI/bFwUCXtibyeZoD```
Screenshot 2024-04-04 at 4 03 13 PM

#### Reproduced on Stellar Laboratory 
The same problem can also be reproduced on the Stellar Laboratory, with the source wallet generated by random Keypair. 
1. Generate a key pair on Stellar Lab and fund the wallet by Friendbot
8. Create a transaction with SetOptions operation that sets 
    - min, medium, max threshold to 1
    - master key weight to 0
    - a signer generated from the Trezor with weight 1 (e.g. from path m/44'/148'/0')
9. Sign the transaction on the Sign Transaction tab with the master key pair secret, and submit the returned signed XDR to the testnet
10. Create another transaction with a operation (it can be any transaction), export its XDR
11. Sign the transaction on the Sign Transaction tab with the path of the Trezor Signer, and submit the returned XDR to the testnet.
12. An error tx_bad_auth will occur

### PR DESCRIPTION
**Pull request by overcat**
fix #3690

### CODE REVIEW

File: legacy/firmware/stellar.c [hunk 1]
```diff
@@ -92,8 +92,12 @@ bool stellar_signingInit(const StellarSignTx *msg) {
   memcpy(&(stellar_activeTx.address_n), &(msg->address_n),
          sizeof(stellar_activeTx.address_n));
 
-  // Hash: public key
-  stellar_hashupdate_address(node->public_key + 1);
```
Comment (overcat) [inline legacy/firmware/stellar.c:96]: Here is the reason for the bug, it should use tx.source.

**Review by matejcik — APPROVED** [review]
all CI failures are expected. please squash the fixup and this is good to go

**Comment by overcat** [comment]
> all CI failures are expected. please squash the fixup and this is good to go

Done!

**Comment by matejcik** [comment]
thanks :100:

**Comment by overcat** [comment]
Hi @matejcik, thank you for your prompt response. 

BTW, when you have plan (and time) to handle PR related to AMM and Soroban, please ping me.

**Pull request merged**